### DISCUSSION TITLE
BME280 issue

### DISCUSSION THREAD

**Discussion by serles** [micropython/micropython#10506] [Using MicroPython]
hello everyone,
somehow my code runs into a problem...

i tried the following exercise: [link](https://how2electronics.com/interface-bme280-with-raspberry-pi-pico-using-micropython/)
And i have bought this BME280: [link](https://www.berrybase.ch/gy-bme280-breakout-board-3in1-sensor-fuer-temperatur-luftfeuchtigkeit-und-luftdruck?gclid=CjwKCAiA5Y6eBhAbEiwA_2ZWIVlhKilpOYpPudLsJYmkZvR1st0ZdYwoXoERtiSaYDoTIO7WFlF0yhoCnGkQAvD_BwE)
CS is pulled up
I used every BME280 library that i found, most of them are looking like the one from the exercise.

I receive the following error:
`Traceback (most recent call last):
  File "<stdin>", line 10, in <module>
  File "/lib/bme280.py", line 74, in __init__
OSError: [Errno 5] EIO`

using an i2c scanner, it seems fine:
`I2C address:
[118]  (decimal)
0x76  (hex)`

Any ideas? :)

**Comment by robert-hh** [comment]
`OSError: [Errno 5] EIO` indicates an I2C communication error. The BME280 library itself is fine and has been used many times. How long are the wires between Pico and Sensor? Check the wiring for bad contacts. Did you try using a lower I2C freq value?

**Comment by serles** [comment]
wires are around 40cm, sensor and wires are plugged on a breadboard. may be cause the issue, i will test today afternoon.
lower i2c freq of 200k has not solved the issue.
good thing is that the i2c scan detects something...
what does the mode like BME280_OSAMPLE_2 exactly do?

**Reply by robert-hh** [reply]
It controls oversampling by internally measuring n times and using the average of n, reducing the noise and increasing the measurement time. You could check the model of the sensor by calling:
```
from machine import SoftI2C, Pin
i2c=SoftI2C(sda=Pin(20), scl=Pin(21), freq=400000) 
print(i2c.readfrom_mem(0x76, 0xd0, 1)[0])
```
I used the Pin setting from the example which you have referenced. I had problems here with Hard I2C, so I used SoftI2C. You may try that as well in your code.
It should return 96, which is the ID of a BME280. If it fails with `OSError: [Errno 5] EIO`, then this is again an indication for a communication problem. If it returns a value, then the basic communication is fine.

**Reply by serles** [reply]
first of all, thanks for the support robert.
It has returned 96.
I replaced the code with softi2c, and it seems to work now, many thanks!
trying to study softi2c i do not fully understand the difference. why has it led to a result?

**Reply by robert-hh** [reply]
When I tried using HW I2C first for confirmation, it failed. That was stunning, because I had used it with these Bosch sensor already often. So I tried SoftI2C, which worked. Then IU hooked up a logic analyzer to see the difference, and the structure of using the HW I2C looked odd. A difference could be the support of clock stretching, which is supp orted by  SoftI2C, and there is issue #8167 reporting problems of the RP2 with clock stretching/timing on short writes, which was solved inJan 2022.
Reloading the firmware with the actual version made HW I2C work again, which pzzled me since I usually use a recent firmware with my boards. You could try the same by uploading today's daily build.

**Reply by serles** [reply]
That was very helpful 
Thanks so much
I would never have tried softi2c
